### PR TITLE
[Messenger] allow Redis transport to claim multiple messages sequentially

### LINKED ISSUE #49023: [Messenger][Redis] Transport can claim only single message per claim_interval
### Symfony version(s) affected

5.4, 6.3

### Description

https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php

After a message is claimed `nextClaim` field is reseted:
https://github.com/symfony/symfony/blob/1f7bc1098f0b03181095d35b11fa48523864e6f2/src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php#L285-L303
which does not allow to claim next messages:
https://github.com/symfony/symfony/blob/1f7bc1098f0b03181095d35b11fa48523864e6f2/src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php#L337-L339

This logic might be intended and works in many cases but it doesn't work well for batch workers with large batch size or large `claim_interval` setting. It can potentially cause messages to pile up in worst cases or to be processed far too late.

### How to reproduce

Possible to reproduce this by SIGKILL'ing a batch worker with messages stored. I hope that my explanation is good enough and a reproducer is not needed.

### Possible Solution

Should be possible to fix this by moving `nextClaim` update to a different place. The field should not be updated if there is a pending message with the same consumer (already does) or a message from another consumer was claimed (this bug).

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by AdamKatzDev**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49023
| License       | MIT

Fix for Redis transport, allows to claim multiple messages sequentially. 
Before the fix a worker could claim only a single message per `claim_interval` from another consumer. In some cases this could lead to problems like processing delays or messages piling up.

### CODE REVIEW

**Comment by AdamKatzDev** [comment]
Not sure why the integration test keeps failing, it seems unrelated to the changes I've made.

**Comment by AdamKatzDev** [comment]
Looks like the tests in AppVeyor are a bit flaky, the code between builds is different only in a comment text:
https://ci.appveyor.com/project/fabpot/symfony/builds/45988646  - RedisExtIntegrationTest::testGetAfterReject crashes
https://ci.appveyor.com/project/fabpot/symfony/builds/45989058  - RedisExtIntegrationTest::testLazy crashes

I suppose that the PR is alright. Though it is a bit suspicious that the Connection tests are failing.

File: src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php [hunk 1]
```diff
@@ -271,57 +287,90 @@ public function testClaimAbandonedMessageWithRaceCondition()
...
+        $redis->expects($this->once())->method('xreadgroup')
+            ->with('symfony', 'consumer', ['queue' => '>'], 1, null) // weren't able to claim the message, asking for new ones
+            ->willReturn([]);
 
-        $redis->expects($this->once())->method('xpending')->willReturn([[
-            0 => 'redisid-123', // message-id
-            1 => 'consumer-2', // consumer-name
-            2 => 3600001, // idle
-        ]]);
+        $redis->expects($this->exactly(2))->method('xpending')->willReturnOnConsecutiveCalls(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php:295]: I forgot to replace willReturnOnConsecutiveCalls with the $series trick (see removed code)

File: src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php [hunk 2]
```diff
@@ -271,57 +287,90 @@ public function testClaimAbandonedMessageWithRaceCondition()
...
+    public function testClaimAbandonedMessageWithRaceConditionExtremeContention()
+    {
+        $redis = $this->createMock(\Redis::class);
+
+        // (xpending -> xclaim) x 3 -> xreadgroup
+        $redis->expects($this->once())->method('xreadgroup')
+            ->with('symfony', 'consumer', ['queue' => '>'], 1, null) // weren't able to claim a message, asking for new ones
             ->willReturn([]);
 
+        $redis->expects($this->exactly(3))->method('xpending')->willReturnOnConsecutiveCalls(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php:321]: here also, willReturnOnConsecutiveCalls should be replaced, PHPUnit deprecated it.

File: src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php [hunk 3]
```diff
@@ -331,17 +380,58 @@ public function testClaimAbandonedMessage()
...
+    {
+        $redis = $this->createMock(\Redis::class);
+
+        // twice: xpending -> xclaim -> xreadgroup
+        $redis->expects($this->exactly(2))->method('xreadgroup')
+            ->withConsecutive(
+                ['symfony', 'consumer', ['queue' => '0'], 1, null], // claim and get message redisid-123
+                ['symfony', 'consumer', ['queue' => '0'], 1, null]  // claim and get message redisid-234
+            )
+            ->willReturnOnConsecutiveCalls(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php:399]: replace

File: src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php [hunk 4]
```diff
@@ -331,17 +380,58 @@ public function testClaimAbandonedMessage()
...
+                ['queue' => [['message' => '{"body":"1","headers":[]}']]],
+                ['queue' => [['message' => '{"body":"2","headers":[]}']]]
+            );
+
+        $redis->expects($this->exactly(2))->method('xpending')->willReturnOnConsecutiveCalls(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ConnectionTest.php:404]: replace

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php [hunk 5]
```diff
@@ -329,13 +325,13 @@ private function claimOldPendingMessages()
                     ['JUSTID']
                 );
 
-                $this->couldHavePendingMessages = true;
+                $couldHavePendingMessages = $couldHavePendingMessages || $claimResult || $try < 3 && $this->claimOldPendingMessages(++$try);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:328]: I added the `$couldHavePendingMessages || ` part, to account for the variable possibly being set to true on L311, does that make sense?

File: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php [hunk 6]
```diff
@@ -308,19 +308,15 @@ private function claimOldPendingMessages()
         $claimableIds = [];
         foreach ($pendingMessages as $pendingMessage) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:309]: on L303, we call
`$this->connection->xpending($this->stream, $this->group, '-', '+', 1);`

Doesn't the `1` means we ask for only one item? Of not, how much iterations can this loop have? If the number is possibly high, do we need to do anything about that?

Sorry if this question is dumb, I'm not into this topic much.

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for the PR. I rebased and made a few changes to it, please fetch before working on your branch. Here are some comments to help move this forward.

**Comment by AdamKatzDev** [comment]
Hi, @nicolas-grekas. Thank you for your time.

The original intent of this PR was to be able to claim multiple stale messages for batch workers, but I've realized that this code won't work for that use case. I've described a possible implementation for new claim algorithm here https://github.com/symfony/symfony/issues/44400#issuecomment-1398477335. But it is also very complicated and introduces a state into the Connection class, so I wasn't confident in the approach.

This PR can improve processing of stale messages when batches are not used, but in that case the queue unlikely to have a lot of stale messages to begin with. I guess the PR simplifies some parts of the code a little bit.

**Comment by nicolas-grekas** [comment]
Thanks for the feedback. Let's close, since this doesn't change the behavior but only the internal implementation, isn't it?

**Comment by AdamKatzDev** [comment]
@nicolas-grekas 

Right.

Redis 6.2.0 introduced IDLE parameter to XPENDING and added XAUTOCLAIM command, it looks like these can help to improve the algorithm significantly, fix the issue that I've tried to tackle here and make the transport usable for batch handlers by also partially fixing https://github.com/symfony/symfony/issues/44400.

I will make another PR if I have some free time.

**Pull request closed without merging**